### ISSUE TITLE
TargetEncoder's out-of-fold encoding leaks through a shared global prior

### ISSUE THREAD

**Issue by bijay-odyssey** [bijay-odyssey/edaprep#36]
`TargetEncoder.fit_transform` is supposed to be leakage-free: "each row is encoded from
the folds that do not contain it -- so a model trained on the output cannot see its own
label through the encoding." The per-fold *counts* really are restricted to the other
folds. The smoothing *prior* mixed into every fold's estimate is not:

```python
# _fit, line 634 -- computed once, over the WHOLE training target
self.prior_ = float(np.nanmean(target)) if np.isfinite(target).any() else 0.0
...
# _fit_transform, per fold, line 672-673
mapping = _smoothed_means(series[rest], target[rest], self.prior_, self.smoothing_)
```

`_smoothed_means` computes `(sum + prior * smoothing) / (count + smoothing)`. `sum` and
`count` are correctly restricted to `rest` (the other folds). `prior` is `self.prior_`,
computed from every row -- including every row in the current holdout fold. For a
category with few or zero occurrences in `rest` (a near-singleton category, exactly the
case this class exists to protect), the encoded value collapses toward `self.prior_`,
which was partly computed from that row's own target.

### Reproduced

```python
df = pd.DataFrame({"cat": ["a", "b", "c", "d", "e"], "y": [0, 0, 0, 0, 0]})
# 5 singleton categories -> every row's OOF value falls back entirely on self.prior_
TargetEncoder(["cat"], n_folds=5, random_state=0).fit_transform(df[["cat"]], df["y"], ctx)["cat"]
# [0.0, 0.0, 0.0, 0.0, 0.0]

df2 = df.copy(); df2["y"] = [1, 0, 0, 0, 0]  # flip ONLY row 0's own label
TargetEncoder(["cat"], n_folds=5, random_state=0).fit_transform(df2[["cat"]], df2["y"], ctx2)["cat"]
# [0.2, 0.2, 0.2, 0.2, 0.2]
```

Row 0's own out-of-fold encoded value changes purely because row 0's own target
changed -- the exact failure mode the docstring says this class prevents.

Why the existing leakage test didn't catch it: `test_target_encoding_is_cross_fitted`
(`tests/test_leakage.py:194`) only asserts an aggregate correlation
(`corr_oof < 0.2`) on a frame where most categories have decent support. The leak is
strongest exactly where support is thinnest -- rare/near-singleton categories -- which a
single aggregate correlation over a whole synthetic frame can easily average away.

### What to do

Compute the prior per fold, from `target[rest]` only, rather than reusing the
already-computed `self.prior_`:

```python
fold_prior = float(np.nanmean(target[rest])) if np.isfinite(target[rest]).any() else self.prior_
mapping = _smoothed_means(series[rest], target[rest], fold_prior, self.smoothing_)
...
out[holdout] = encoded.fillna(fold_prior).to_numpy(dtype=np.float64)
```

`self.prior_` (the full-train prior) stays correct and necessary for `_transform` on
genuinely new data at prediction time -- it's only the `_fit_transform` per-fold path
that needs its own, fold-restricted prior.

### Tests

- Extend or add alongside `test_target_encoding_is_cross_fitted`: a frame where every
  category is a near-singleton (as in the repro above), asserting that flipping one
  row's own target does not change that row's own OOF-encoded value.
- `test_target_encoder_fit_then_transform_differs_from_fit_transform` already exists;
  worth checking it still passes and still demonstrates a real difference after the fix.

**Before you start:** please comment to claim this issue, and check whether someone
already has.

**Comment by XonkelX** [comment]
Hi! I’d like to work on this issue. I reviewed the leakage path in `TargetEncoder.fit_transform()`, and the per-fold prior approach described here makes sense to me.

My plan is to compute the smoothing prior only from each fold’s training partition, keep the full-training prior unchanged for normal `transform()`, and add a regression test with near-singleton categories to verify that changing a row’s own target no longer affects its OOF encoding.

If no one else is working on it, I’d be happy to take this one.

**Comment by bijay-odyssey** [comment]
Assigned — and this is exactly the right fix. Two things worth flagging before you start, since they're easy to miss:

1. **The `.fillna()` on the holdout side needs the fold prior too.** `_fit_transform` does `encoded = series[holdout].map(mapping); out[holdout] = encoded.fillna(self.prior_)...` — for a category entirely absent from `rest` (the exact near-singleton case this bug bites hardest), `mapping` has no entry for it at all, so the row falls through to that `.fillna(self.prior_)` rather than through `_smoothed_means`'s prior term. If that stays `self.prior_` instead of your fold-restricted prior, you'll fix the partial-leak case and leave the "category entirely missing from this fold" case still leaking 100%, which is worse. Both uses in the fold loop need the same `fold_prior`.

2. **Edge case to decide on purpose, not by accident:** what happens when a fold's `rest` has zero finite target values at all (possible with a tiny frame and enough folds)? `np.nanmean` on an all-NaN slice warns and returns `nan`. Falling back to `self.prior_` in that specific situation seems reasonable to me (there's nothing fold-local to compute), but say so explicitly in a comment rather than let it happen implicitly — this codebase has been bitten before by an implicit fallback nobody wrote down (that's most of what #18 was about).

Your test plan is the right one. If you want a stronger version: assert not just that flipping one row's target doesn't change its own OOF value, but that it doesn't change *any other row's* OOF value either in the singleton case — a shared `self.prior_` leaking would move every row at once, which is actually easier to catch than it sounds and worth having explicitly.
